src/js/protocols/webstm32.js

```diff
@@ -0,0 +1,941 @@
+                                onDisconnect(true);
+                            }, 100);
+                        };
+
+                        function onAbort() {
+                            GUI.connect_lock = false;
+                            rebootMode = 0;
+                            console.log('User cancelled because selected target does not match verified board');
+                            reboot();
+                            TABS.firmware_flasher.refresh();
+                        }
+                        try {
+                            if (selectedBoard !== connectedBoard && !TABS.firmware_flasher.localFirmwareLoaded) {
+                                TABS.firmware_flasher.showDialogVerifyBoard(selectedBoard, connectedBoard, onAbort, reboot);
+                            } else {
+                                reboot();
+                            }
+                        } catch (e) {
+                            console.error(e);
+                            reboot();
+                        }
+                    });
+                }
+            }
+
+            const onTimeoutHandler = function () {
+                GUI.connect_lock = false;
+                console.log('Looking for capabilities via MSP failed');
+
+                TABS.firmware_flasher.flashingMessage(i18n.getMessage('stm32RebootingToBootloaderFailed'), TABS.firmware_flasher.FLASH_MESSAGE_TYPES.INVALID);
+            };
+
+            const onFailureHandler = function () {
+                GUI.connect_lock = false;
+                TABS.firmware_flasher.refresh();
+            };
+
+            GUI.connect_lock = true;
+            TABS.firmware_flasher.flashingMessage(i18n.getMessage('stm32RebootingToBootloader'), TABS.firmware_flasher.FLASH_MESSAGE_TYPES.NEUTRAL);
+
+            this.msp_connector.connect(this.port, this.options.reboot_baud, onConnectHandler, onTimeoutHandler, onFailureHandler);
+        }
+    }
+    // initialize certain variables and start timers that oversee the communication
+    initialize() {
+
+        console.log(":exploding_head:");
+
+        serial.removeEventListener('receive', read_serial_adapter);
+        serial.addEventListener('receive', read_serial_adapter);
+
+        // reset and set some variables before we start
+        this.receive_buffer = [];
+        this.verify_hex = [];
+
+        this.upload_time_start = new Date().getTime();
+        this.upload_process_alive = false;
+
+        // reset progress bar to initial state
+        TABS.firmware_flasher.flashingMessage(null, TABS.firmware_flasher.FLASH_MESSAGE_TYPES.NEUTRAL)
+            .flashProgress(0);
+
+        // lock some UI elements TODO needs rework
+        $('select[name="release"]').prop('disabled', true);
+
+        // TODO: update to use web serial / USB API
+        serial.onReceive.addListener((info) => {
+            this.read(info);
+        });
```

Comment:
I don't think we even have `onReceive` in web serial implementation. We are using standard event target there.

Reply:
Yes that part is still open. Fixed anyway.